### PR TITLE
FAI-273 - make PDP work with any model

### PR DESCRIPTION
**Pull request by tteofili**
This PR modifies the `PartialDependencePlotExplainer` so that it can be used with any kind of model, not just models relying on numeric features only (as it is up to now).

See https://issues.redhat.com/browse/FAI-273

Many thanks for submitting your Pull Request :heart:! 

Please make sure that your PR meets the following requirements:

- [x] You have read the [contributors guide](https://github.com/kiegroup/kogito-runtimes#contributing-to-kogito)
- [ ] Pull Request title is properly formatted: `KOGITO-XYZ Subject`
- [x] Pull Request title contains the target branch if not targeting master: `[0.9.x] KOGITO-XYZ Subject`
- [x] Pull Request contains link to the JIRA issue
- [x] Pull Request contains link to any dependent or related Pull Request
- [x] Pull Request contains description of the issue
- [x] Pull Request does not include fixes for issues other than the main ticket

### CODE REVIEW

File: explainability/explainability-core/src/main/java/org/kie/kogito/explainability/global/pdp/PartialDependencePlotExplainer.java [hunk 1]
```diff
@@ -15,35 +15,20 @@
...
-import org.kie.kogito.explainability.model.PredictionOutput;
-import org.kie.kogito.explainability.model.PredictionProvider;
-import org.kie.kogito.explainability.model.PredictionProviderMetadata;
-import org.kie.kogito.explainability.model.Value;
-import org.kie.kogito.explainability.utils.DataUtils;
+import org.kie.kogito.explainability.model.*;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
+import java.util.*;
```
Comment (danielezonca) [inline explainability/explainability-core/src/main/java/org/kie/kogito/explainability/global/pdp/PartialDependencePlotExplainer.java]: Please use explicit import

File: explainability/explainability-core/src/main/java/org/kie/kogito/explainability/global/pdp/PartialDependencePlotExplainer.java [hunk 2]
```diff
@@ -136,6 +109,67 @@ public PartialDependencePlotExplainer() {
...
+                try {
+                    Map<String, Long> classCount = (Map<String, Long>) value.getUnderlyingObject();
+                    if (classCount.containsKey(categoricalOutput)) {
+                        classCount.put(categoricalOutput, classCount.get(categoricalOutput) + 1);
+                    } else {
+                        classCount.put(categoricalOutput, 1L);
+                    }
+                    marginalImpacts.set(i, new Value<>(classCount));
+                } catch (ClassCastException cce) {
+                    // ignore malformed output
```
Comment (danielezonca) [inline explainability/explainability-core/src/main/java/org/kie/kogito/explainability/global/pdp/PartialDependencePlotExplainer.java]: Please log error (don't use jboss logmanager! :P )

File: explainability/explainability-integrationtests/explainability-integrationtests-opennlp/src/test/java/org/kie/kogito/explainability/explainability/integrationtests/opennlp/OpenNLPPDPExplainerTest.java [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+import opennlp.tools.langdetect.LanguageDetector;
+import opennlp.tools.langdetect.LanguageDetectorME;
+import opennlp.tools.langdetect.LanguageDetectorModel;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.Test;
+import org.kie.kogito.explainability.Config;
+import org.kie.kogito.explainability.global.pdp.PartialDependencePlotExplainer;
+import org.kie.kogito.explainability.local.lime.LimeConfig;
+import org.kie.kogito.explainability.local.lime.LimeExplainer;
+import org.kie.kogito.explainability.model.*;
```
Comment (danielezonca) [inline explainability/explainability-integrationtests/explainability-integrationtests-opennlp/src/test/java/org/kie/kogito/explainability/explainability/integrationtests/opennlp/OpenNLPPDPExplainerTest.java]: Explicit imports

File: explainability/explainability-integrationtests/explainability-integrationtests-dmn/src/test/java/org/kie/kogito/explainability/explainability/integrationtests/dmn/DecisionModelWrapper.java [hunk 4]
```diff
@@ -56,7 +56,14 @@
             DMNResult dmnResult = decisionModel.evaluateAll(context);
             List<Output> outputs = new LinkedList<>();
             for (DMNDecisionResult decisionResult : dmnResult.getDecisionResults()) {
-                Output output = new Output(decisionResult.getDecisionName(), Type.TEXT, new Value<>(decisionResult.getResult()), 1d);
+                Value<Object> value = new Value<>(decisionResult.getResult());
+                Type type;
+                if (Double.isNaN(value.asNumber())) {
```
Comment (danielezonca) [inline explainability/explainability-integrationtests/explainability-integrationtests-dmn/src/test/java/org/kie/kogito/explainability/explainability/integrationtests/dmn/DecisionModelWrapper.java:61]: Why this check? What about other types?

Comment (tteofili) [reply explainability/explainability-integrationtests/explainability-integrationtests-dmn/src/test/java/org/kie/kogito/explainability/explainability/integrationtests/dmn/DecisionModelWrapper.java:61]: this is just to differentiate between numeric outputs and _all other types_ in the tests.
The reason why it's there is that the `PDPExplainer` differentiates between numeric and non-numeric outputs.
In practice, actual DMN models returning the correct `Type` would be a responsibility of the `explainability-service` or eventually of the `explainability-core` correctly parsing the output value; `DecisionModelWrapper` is just used for testing purposes so I had thought checking all `Types` wasn't needed.

I can anyway delegate the validation responsibility to the `PDPExplainer` too, to make sure that `Types` and `Values` are parsed/handled correctly.

File: explainability/explainability-core/src/main/java/org/kie/kogito/explainability/global/pdp/PartialDependencePlotExplainer.java (hunk 2 above)
Comment (tteofili) [reply explainability/explainability-core/src/main/java/org/kie/kogito/explainability/global/pdp/PartialDependencePlotExplainer.java]: ok :)

File: explainability/explainability-integrationtests/explainability-integrationtests-opennlp/src/test/java/org/kie/kogito/explainability/explainability/integrationtests/opennlp/OpenNLPPDPExplainerTest.java (hunk 3 above)
Comment (tteofili) [reply explainability/explainability-integrationtests/explainability-integrationtests-opennlp/src/test/java/org/kie/kogito/explainability/explainability/integrationtests/opennlp/OpenNLPPDPExplainerTest.java]: ok

File: explainability/explainability-core/src/main/java/org/kie/kogito/explainability/global/pdp/PartialDependencePlotExplainer.java (hunk 1 above)
Comment (tteofili) [reply explainability/explainability-core/src/main/java/org/kie/kogito/explainability/global/pdp/PartialDependencePlotExplainer.java]: ok

File: explainability/explainability-integrationtests/explainability-integrationtests-dmn/src/test/java/org/kie/kogito/explainability/explainability/integrationtests/dmn/DecisionModelWrapper.java (hunk 4 above)
Comment (danielezonca) [reply explainability/explainability-integrationtests/explainability-integrationtests-dmn/src/test/java/org/kie/kogito/explainability/explainability/integrationtests/dmn/DecisionModelWrapper.java:61]: Well if this is a corner case we can add a log message too :thinking:

Comment (tteofili) [reply explainability/explainability-integrationtests/explainability-integrationtests-dmn/src/test/java/org/kie/kogito/explainability/explainability/integrationtests/dmn/DecisionModelWrapper.java:61]: correct, will do it

File: explainability/explainability-integrationtests/explainability-integrationtests-opennlp/src/test/java/org/kie/kogito/explainability/explainability/integrationtests/opennlp/OpenNLPPDPExplainerTest.java [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+import org.kie.kogito.explainability.model.PartialDependenceGraph;
+import org.kie.kogito.explainability.model.Prediction;
+import org.kie.kogito.explainability.model.PredictionInput;
+import org.kie.kogito.explainability.model.PredictionOutput;
+import org.kie.kogito.explainability.model.PredictionProvider;
+import org.kie.kogito.explainability.model.Type;
+import org.kie.kogito.explainability.model.Value;
+
+import java.io.InputStream;
+import java.util.*;
```
Comment (danielezonca) [inline explainability/explainability-integrationtests/explainability-integrationtests-opennlp/src/test/java/org/kie/kogito/explainability/explainability/integrationtests/opennlp/OpenNLPPDPExplainerTest.java]: Import to fix

File: explainability/explainability-core/src/main/java/org/kie/kogito/explainability/global/pdp/PartialDependencePlotExplainer.java (hunk 1 above)
Comment (danielezonca) [reply explainability/explainability-core/src/main/java/org/kie/kogito/explainability/global/pdp/PartialDependencePlotExplainer.java]: Please review because it is not solved

File: explainability/explainability-integrationtests/explainability-integrationtests-opennlp/src/test/java/org/kie/kogito/explainability/explainability/integrationtests/opennlp/OpenNLPPDPExplainerTest.java (hunk 5 above)
Comment (tteofili) [reply explainability/explainability-integrationtests/explainability-integrationtests-opennlp/src/test/java/org/kie/kogito/explainability/explainability/integrationtests/opennlp/OpenNLPPDPExplainerTest.java]: 👍

**Comment by tteofili** [comment]
Schermata 2021-01-28 alle 14 09 27

Sample `PartialDependenceGraph` generated with `PartialDependenceExplainer` on `TrafficViolationDMN`, correlating `speed` feature with `Fine` output.

File: explainability/explainability-core/src/main/java/org/kie/kogito/explainability/global/pdp/PartialDependencePlotExplainer.java [hunk 6]
```diff
@@ -136,6 +126,90 @@ public PartialDependencePlotExplainer() {
...
+                        output = entry.getKey().asString();
+                    }
+                }
+                yValues.add(new Value<>(output));
+            }
+        }
+        return yValues;
+    }
+
+    private void updateValueCounts(List<Map<Value<?>, Long>> valueCounts, int i, Output output) {
```
Comment (danielezonca) [inline explainability/explainability-core/src/main/java/org/kie/kogito/explainability/global/pdp/PartialDependencePlotExplainer.java]: Please use a different variable name instead of `i` :)

File: explainability/explainability-core/src/main/java/org/kie/kogito/explainability/utils/DataUtils.java [hunk 7]
```diff
@@ -433,4 +439,24 @@ private static void linearizeFeature(List<Feature> flattenedFeatures, Feature f)
...
+            List<Value<?>> xAxis = partialDependenceGraph.getX();
+            List<Value<?>> yAxis = partialDependenceGraph.getY();
+            outputStream.write("feature,output\n".getBytes(StandardCharsets.UTF_8));
+            for (int i = 0; i < xAxis.size(); i++) {
+                String line = xAxis.get(i).asString().replaceAll(",", "") + ',' +
+                        yAxis.get(i).asString().replaceAll(",", "") + '\n';
+                outputStream.write(line.getBytes(StandardCharsets.UTF_8));
+            }
+            outputStream.flush();
+        }
```
Comment (danielezonca) [inline explainability/explainability-core/src/main/java/org/kie/kogito/explainability/utils/DataUtils.java:460]: What about use [`apache-commons-csv`](https://commons.apache.org/proper/commons-csv/)?

**Review by danielezonca — APPROVED** [review]
Two minor comments, approving :+1:

File: explainability/explainability-core/src/main/java/org/kie/kogito/explainability/utils/DataUtils.java (hunk 7 above)
Comment (tteofili) [reply explainability/explainability-core/src/main/java/org/kie/kogito/explainability/utils/DataUtils.java:460]: I've created https://issues.redhat.com/browse/FAI-383 to follow up on this (it'd require updating / changing the `kogito-bom` in a separate PR so I think it's better to do that in a separate issue / PR).
